### DISCUSSION TITLE
Nullability enforcements when working with Deconstruct

### DISCUSSION THREAD

**Discussion by AlbertoMonteiro** [dotnet/csharplang#4804] [Q&A]
In NET 5 we have many attributes that help the compiler to know when a field may be null or not.

But it lacks support when you use Deconstruct.

I have the given type

```csharp
public struct Result
{
    public Exception? Exception { get; }

    [MemberNotNullWhen(false, nameof(Exception))]
    public bool IsSuccess { get; }

    public Result(bool success)
    {
        IsSuccess = success;
        Exception = null;
    }

    public Result(Exception exception)
    {
        Exception = exception;
        IsSuccess = false;
    }

    public bool ErrorIs<TException>()
        where TException : Exception
        => Exception is TException;

    public static Result Success()
        => new(true);

    public static Result Error(Exception exception)
        => new(exception);

    public static Result<T> Success<T>(T value)
        => new(value);

    public static Result<T> Error<T>(Exception exception)
        => new(exception);

    public static implicit operator Result(Exception exception)
        => new(exception);

    public static implicit operator bool(Result result)
        => result.IsSuccess;

    public static implicit operator Task<Result>(Result result)
        => result.AsTask;

    public void Deconstruct(out bool success, out Exception? exception)
        => (success, exception) = (IsSuccess, Exception);

    public Task<Result> AsTask => Task.FromResult(this);
}
```

When I use it, the compiler gives me the proper warning if I do this:

![image](https://user-images.githubusercontent.com/836496/120846818-0a3d3500-c549-11eb-9f4e-e1c8eb0bd0b3.png)

No warning, since IsSuccess is false

![image](https://user-images.githubusercontent.com/836496/120846872-1a551480-c549-11eb-90ed-5f549af2f6a0.png)

But this `Result` type supports Deconstruct, and when I use it, I don't have the same benefit of this previous version

![image](https://user-images.githubusercontent.com/836496/120847039-4f616700-c549-11eb-991f-8ef2c4049385.png)

Also, I cant use the `NotNullWhen`, since the method does not return

```csharp
public void Deconstruct(out bool success, [NotNullWhen(false)] out Exception? exception)
    => (success, exception) = (IsSuccess, Exception);
```

Is there a way to do it, and if not, are there any plans to support that kind of feature?

**Comment by RikkiGibson** [comment]
I think it's reasonable that people want this, but that if we chose to support it, that it might end up being part of a general "associate flow state with bool locals" feature. e.g.

```cs
object? x = ...;
bool isNull = x == null;
if (!isNull) x.ToString(); // ok
```

since it seems to exhibit the same problems of:
1. how do you store conditional state associated with a bool local "on the side" that gets integrated in when the bool is tested
2. how do you invalidate that state appropriately when effects occur in-between the assignment and usage of the local e.g.
```cs
(Success, Exception) = Result.Success();
Exception = null; // should invalidate conditional state associated with 'Success'
if (!Success) Exception.ToString(); // should warn
```

**Reply by 333fred** [reply]
We would need a general alias analysis to perform this, which we have previously said is not worth the tradeoffs in performance.

**Reply by AlbertoMonteiro** [reply]
@RikkiGibson this already happens today.
![image](https://user-images.githubusercontent.com/836496/121616520-70d6bd00-ca39-11eb-8d2d-1bde48927fe4.png)
As you can see there is no warning about the exception being null.

Because I annotated like that 
![image](https://user-images.githubusercontent.com/836496/121616581-864be700-ca39-11eb-83b9-fed552e80735.png)

And if I change the if statement, then it warns. 
![image](https://user-images.githubusercontent.com/836496/121616703-c4e1a180-ca39-11eb-8dd4-8eff9cbf1858.png)

So, if we had the possibilitity to do that in deconstructing way, we would be in same "place" as using `MemberNotNullWhen`, it would be partially covered.
Better than nothing.

**Reply by 333fred** [reply]
This is one of those cases described in https://xkcd.com/1425: it sounds easy, but is actually quite hard. Properties are a single location where the access occurs _and_ the check occurs at the same location. Doing the same thing where one local location affects another local location is significantly more complex and, while possible, has generally not met the team's cost/benefit ratio for both the implementation difficulty and the cost to runtime performance.

**Reply by AlbertoMonteiro** [reply]
Gotcha @333fred ,thanks for clarification
